src/linker/Linker/LinkerAttributesInformation.cs

```diff
@@ -12,42 +13,63 @@ namespace Mono.Linker
 {
 	readonly struct LinkerAttributesInformation
 	{
-		readonly Dictionary<Type, List<Attribute>> _linkerAttributes;
+		readonly Dictionary<Type, List<object>> _linkerAttributes;
 
-		public LinkerAttributesInformation (LinkContext context, MethodDefinition method)
+		public LinkerAttributesInformation (LinkContext context, IMemberDefinition member)
 		{
 			_linkerAttributes = null;
 
-			if (method.HasCustomAttributes) {
-				foreach (var customAttribute in method.CustomAttributes) {
+			if (member.HasCustomAttributes) {
+				foreach (var customAttribute in member.CustomAttributes) {
 					var attributeType = customAttribute.AttributeType;
-					Attribute attributeValue = null;
-					if (attributeType.Name == "RequiresUnreferencedCodeAttribute" && attributeType.Namespace == "System.Diagnostics.CodeAnalysis") {
-						attributeValue = ProcessRequiresUnreferencedCodeAttribute (context, method, customAttribute);
-					}
-
-					if (attributeValue != null) {
-						if (_linkerAttributes == null)
-							_linkerAttributes = new Dictionary<Type, List<Attribute>> ();
+					object attributeValue = null;
+					if (IsAttribute<RequiresUnreferencedCodeAttribute> (attributeType))
+						attributeValue = ProcessRequiresUnreferencedCodeAttribute (context, member, customAttribute);
+					else if (IsAttribute<DynamicDependencyAttribute> (attributeType))
+						attributeValue = ProcessDynamicDependencyAttribute (context, member, customAttribute);
+					AddAttribute (ref _linkerAttributes, attributeValue);
+				}
+			}
+		}
 
-						Type attributeValueType = attributeValue.GetType ();
-						if (!_linkerAttributes.TryGetValue (attributeValueType, out var attributeList)) {
-							attributeList = new List<Attribute> ();
-							_linkerAttributes.Add (attributeValueType, attributeList);
-						}
+		public LinkerAttributesInformation (LinkContext context, FieldDefinition field)
```

Comment:
As noted above - I would simply have this based on ICustomAttributeProvider - and no special code for different types...

Reply:
Changed to `IMemberDefinition` which implements `ICustomAttributeProvider`. If we need to use the cache for attributes on non-members, we can always relax it to `ICustomAttributeProvider`.